### ISSUE TITLE
false positive: Performance/HashEachMethods

### ISSUE THREAD

**Issue by dennisvandehoef** [bbatsov/rubocop#5328]
Performance/HashEachMethods raises on all `keys.each`, also it should look for `.keys.each` to prevent a false positive like this:

```ruby
def foobar(keys = [])
  keys.each { | key| puts key }
end
```

```
rubocop -v
0.52.0
```

**Comment by bbatsov** [comment]
> Performance/HashEachMethods raises on all keys.each, also it should look for .keys.each to prevent a false positive like this:

Yeah, probably we should check for an explicit receiver to reduce the number of false positives.

**Comment by ragesoss** [comment]
I hit another kind of false positive with this cop:

```
Performance/HashEachMethods: Use each_value instead of each.
    sub_data.each do |_a_id, article_data|
```

In this case, `.each` is being called on an array of two-item arrays, so the block arguments are destructuring the array, like this:

```ruby
sub_data = [ [1,2], [3,4] ]
sub_data.each do |_x, y|
  # do something with y
end
```

**Comment by ragesoss** [comment]
Ah, I see there's a helpful note in the source for my use case"

```
      # Note: If you have an array of two-element arrays, you can put
      #   parentheses around the block arguments to indicate that you're not
      #   working with a hash, and suppress RuboCop offenses.
```

Maybe adding something along these lines to the cop output would be helpful?

**Comment by koic** [comment]
Hi @dennisvandehoef.

I could not reproduce a false positive at RuboCop 0.52.0 and 0.52.1.

```console
% rubocop -v
0.52.1
```

```console
% cat /tmp/example.rb
def foobar(keys = [])
  keys.each { | key| puts key }
end
```

```console
% rubocop /tmp/example.rb --only Performance/HashEachMethods
Inspecting 1 file
.

1 file inspected, no offenses detected
```

Would you please show me reproduction steps? Or is there a mistake in my understanding?

**Comment by reitzig** [comment]
Similar:

```ruby
class Test
  def keys
    [1,2]
  end
end

Test.new.keys.each { |s| puts s }
```

RubuCop proposes the change here, even though the receiver is no Hash (and does indeed not have `each_key`).

**Comment by Drenmi** [comment]
This cop is removed in the next release, as the performance improvements are no longer present in current Ruby.
